ui/raidboss/emulator/data/PopupTextAnalysis.js

```diff
@@ -27,29 +75,37 @@ export default class PopupTextAnalysis extends StubbedPopupText {
       for (const trigger of this.triggers) {
         const r = log.match(trigger.localRegex);
         if (r) {
-          // Some async magic here, force waiting for the entirety of
-          // the trigger execution before continuing
-          const delayPromise = new Promise((res) => {
-            this.delayResolver = res;
+          const resolver = this.currentResolver = new Resolver();
+          resolver.status = {
+            initialData: EmulatorCommon.cloneData(this.data),
+            condition: undefined,
+            response: false,
+            result: false,
+            delay: false,
+            suppressed: false,
+            executed: false,
+            promise: undefined,
+          };
+          this.currentTriggerStatus = resolver.status;
+          this.triggerResolvers.push(resolver);
+
+          this.OnTrigger(trigger, r, logObj.timestamp);
+
+          resolver.setFinal(() => {
+            resolver.status.finalData = EmulatorCommon.cloneData(this.data);
+            delete resolver.triggerHelper.resolver;
+            this.callback(logObj, resolver.triggerHelper, resolver.status, this.data);
           });
-          const promisePromise = new Promise((res) => {
-            this.promiseResolver = res;
-          });
-          const runPromise = new Promise((res) => {
-            this.runResolver = res;
-          });
-
-          this.OnTrigger(trigger, r);
-
-          await delayPromise;
-          await promisePromise;
-          const triggerHelper = await runPromise;
-
-          this.currentTriggerStatus.finalData = EmulatorCommon.cloneData(this.data);
-
-          this.callback(logObj, triggerHelper, this.currentTriggerStatus);
         }
       }
+
+      for (let i = 0; i < this.triggerResolvers.length; ++i) {
+        const ret = await this.triggerResolvers[i].IsResolved(logObj);
+        if (ret)
+          this.triggerResolvers[i] = false;
+      }
+
+      this.triggerResolvers = this.triggerResolvers.filter((resolver) => resolver);
```

Comment:
Can't use filter directly because we can't use `await` in the filter function.

The for loop could alter the array in-line instead but I felt that was a worse solution?

Reply:
For reference, the error that results if trying to use filter directly:
![image](https://user-images.githubusercontent.com/6119598/118376012-70aae500-b593-11eb-8b58-beadcd926d16.png)

And can't pass an async function to filter because those return a Promise which means nothing gets filtered out.

Reply:
https://advancedweb.hu/how-to-use-async-functions-with-array-filter-in-javascript/ has a one-liner suggestion using `Promise.all` to map everything to a promise and then await them all.

Reply:
Swapped to this setup, which works:
```js
  async checkResolved(logObj) {
    await Promise.all(
        this.triggerResolvers.map(async (resolver) => await resolver.isResolved(logObj)))
      .then((results) => {
        this.triggerResolvers = this.triggerResolvers.filter((_, index) => !results[index]);
      });
  }
```

Any additional adjustments to this, or is it good for the next commit?

Reply:
Looks good to me, thanks! It just seems better from a TypeScript/functional programming pov to not reuse the triggerResolvers entries.